org.spoofax.terms/src/org/spoofax/terms/io/binary/SAFWriter.java

```diff
@@ -56,581 +56,498 @@
 import org.spoofax.interpreter.terms.IStrategoString;
 import org.spoofax.interpreter.terms.IStrategoTerm;
 import org.spoofax.interpreter.terms.IStrategoTuple;
+import org.spoofax.terms.io.TermWriter;
 import org.spoofax.terms.util.TermUtils;
 
 /**
- * Writes the given ATerm to a (streamable) binary format. Supply the
```

Comment:
I'm just going to ignore this absolutely horrible diff and assume you didn't mess with the internals of this class. I think the diff is this bad because of reordering of methods? Please be mindful of that in the future, because I basically cannot review this properly.

Reply:
This is very bad, as the functionality of this class is used in many places and small changes can have large consequences, which maybe only be noticed later due to lack of tests and lack of proper bootstrapping. It is very important to be able to review these kind of changes. Please don't refactor everything that you have to touch, but make reviewable changes. If not possible, either not change them or first discuss.

Reply:
I hardly changed anything. The diff is just messed up.

Reply:
New attempt: I verbatim copied the original `SAFWriter` into my code and made the minimal changes. The diff looks better now, but the methods are in weird places and the whitespace is off when you open it in an editor.

Reply:
Let me explain what the refactoring is: the original `SAFWriter` implementation is now a static nested class, but the implementation has not been touched. The `TermWriter` interface's `write(IStrategoTerm, OutputStream)` method does exactly the same the original static `writeTermToSAFStream()` method did. The other methods `writeTermToSAFString()` and `writeTermToSAFFile()` are already covered by the default interface implementation in `TermWriter` and therefore they have been deprecated and their bodies have been replaced by calls to the `TermWriter` implementation.

Reply:
Ah, because of the nested class. You replaced `writeTermToSAFFile` and `writeTermToSAFString` with a call to `write` (`writeTermToSAFStream`), are you sure that does not change anything, because they seem to be doing slightly different things?

Reply:
Basically, what each of these methods do is: serialize the term to the byte buffer, append two bytes that indicate the size of the block that was written (16-bit little-endian), and write the result to the output. `writeTermToSAFFile` uses two byte buffers, one for the serialized data and one two-byte buffer for writing the extra block size, but the result is the same. `writeTermToSAFString` creates a new byte buffer every time, and modifies the resulting array of bytes instead to insert the block size, but the result is the same.

Reply:
I verified this with tests.